survivalstan/formulas.py

```diff
@@ -0,0 +1,315 @@
+    time_df = time_df.join(t_durs)
+    if len(time_df.index)>1:
+        time_df.fillna(inplace=True, value=time_df.loc[1, timepoint_end_col])
+    return(time_df)
+
+
+class Id(object):
+    def __init__(self, desc='id'):
+        self.values = []
+        self.desc = desc
+
+    def memorize_chunk(self, x):
+        self.values.extend(np.unique(x))
+
+    def memorize_finish(self):
+        self.ids = np.arange(len(self.values))+1
+        self.lookup = dict(zip(self.values, self.ids))
+
+    def transform(self, x):
+        if patsy.util.have_pandas and isinstance(x, pd.Series):
+            d = pd.Series([self.lookup[val] for val in x]).astype(int)
+            d.index = x.index
+            return(d)
+        else:
+            return np.array([self.lookup[val] for val in x])
+
+    def len(self):
+        return len(self.ids)
+
+    def decode_df(self):
+        df = pd.DataFrame({'id': self.ids, 'value': self.values})
+        df.dropna(inplace=True)
+        return df
+
+as_id = patsy.stateful_transform(Id)
+
+class SurvData(pd.DataFrame):
+    ''' patsy.DesignMatrix representing survival data output '''
+    survival_type = 'wide'
+
+    def __init__(self, *args, **kwargs):
+        my_kwargs = dict(**kwargs)
+        if 'stan_data' in my_kwargs.keys():
+            stan_data = my_kwargs['stan_data']
+            del my_kwargs['stan_data']
+        else:
+            stan_data = dict()
+        if 'meta_data' in my_kwargs.keys():
+            meta_data = my_kwargs['meta_data']
+            del my_kwargs['meta_data']
+        else:
+            meta_data = dict()
+        super(SurvData, self).__init__(*args, **my_kwargs)
+        self.stan_data = stan_data
+        self.meta_data = meta_data
+
+class LongSurvData(SurvData):
+    ''' pd.DataFrame representing survival data with endpoint_time_id, event_status & subject_id '''
+    survival_type = 'long'
+
+class NotValidId(ValueError):
+    ''' Class of errors pertaining to invalid Id variables '''
+
+class Surv(object):
+    ''' Class representing stateful Survival-type data
+    '''
+    def __init__(self):
+        self.subject_id = Id('subject')
+        self.timepoint_id = Id('timepoint')
+        self.group_id = Id('group')
+        self._type = None
+        self._grouped = None
+        pass
+
+    def _check_kwargs(self, **kwargs):
+        kwargs = dict(**kwargs)
+        allowed_kwargs = ['subject', 'group']
+        bad_keys = [key not in allowed_kwargs for key in kwargs.keys()]
+        if any(bad_keys):
+            raise ValueError('Invalid parameter: {}'.format(','.join(bad_keys)))
+        return kwargs
+
+    def memorize_chunk(self, time, event_status, **kwargs):
+        kwargs = self._check_kwargs(**kwargs)
+        if 'subject' in kwargs.keys():
+            self._type = 'long'
+            self.subject_id.memorize_chunk(kwargs['subject'])
+            self.timepoint_id.memorize_chunk(time)
+        else:
+            self._type = 'wide'
```

Comment:
This is a good point - I should probably review these to see how many are necessary. One of the goals of adding the typed dataframes here is to eventually get away from having to do so many checks. One of my goals was to make it so that dataframe could have its own stan_data dictionary attached, so that the downstream users of that dataframe could just use it without needing to check to confirm wide/long status.